### PR TITLE
refactor: use require.resolve instead of fs.exists

### PR DESCRIPTION
**Pull request by atian25**
##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Description of change

- https://github.com/eggjs/egg-core/pull/237#discussion_r780303089
- https://github.com/eggjs/egg/issues/4970

### CODE REVIEW

**Review by hyj1991 — APPROVED** [review]

**Comment by atian25** [comment]
如 test/fixtures 里面，如果上层的单测的 fixtures 里面有手动 mock 的 node_modules ，对应的 plugin 没有 package.json 则都会单测挂掉，需要手动加个 package.json

**Comment by atian25** [comment]
单测写起来比较恶心，需要手动模拟 pnpm 的目录结构，只模拟了真实文件的结构，没有做软链。

**Review by fengmk2 — APPROVED** [review]

File: lib/loader/mixin/plugin.js [hunk 1]
```diff
@@ -334,39 +334,31 @@ module.exports = {
...
       // 'node_modules/.pnpm/egg@2.33.1/node_modules', <- this is the sibling directory
-      lookupDirs.add(path.dirname(eggPath));
-    }
-
-    for (let dir of lookupDirs) {
-      dir = path.join(dir, name);
-      if (fs.existsSync(dir)) {
-        return fs.realpathSync(dir);
-      }
+      const filePath = require.resolve(`${name}/package.json`, { paths: [ ...lookupDirs ] });
```
Comment (atian25) [inline lib/loader/mixin/plugin.js:357]: 第二个属性要求 8.9 以上版本
https://nodejs.org/dist/latest/docs/api/modules.html#requireresolverequest-options
image

Comment (mansonchor) [reply lib/loader/mixin/plugin.js:357]: > with the exception of [GLOBAL_FOLDERS](https://nodejs.org/dist/latest/docs/api/modules.html#loading-from-the-global-folders) like $HOME/.node_modules, which are always included

`require.resolve` 会最优先找 `$HOME/.node_modules`，和之前表现不一致了，

Comment (atian25) [reply lib/loader/mixin/plugin.js:357]: 应该是我们 layer 的使用方式导致的，我先临时 dist-tag 回滚回去，我们内部修复用法后再更新回去

Comment (atian25) [reply lib/loader/mixin/plugin.js:357]: https://github.com/eggjs/egg-core/pull/246

**Pull request merged**